server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java

```diff
@@ -8,172 +8,166 @@
 
 package org.elasticsearch.action.admin.cluster.stats;
 
-import org.HdrHistogram.DoubleHistogram;
+import org.elasticsearch.common.util.Maps;
 
-import java.util.HashMap;
+import java.util.Collections;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.LongAdder;
 
 /**
  * Service holding accumulated CCS search usage statistics. Individual cross-cluster searches will pass
  * CCSUsage data here to have it collated and aggregated. Snapshots of the current CCS Telemetry Usage
- * can be obtained by getting CCSTelemetrySnapshot objects (TODO: create that class (and determine the best name)).
+ * can be obtained by getting {@link CCSTelemetrySnapshot} objects.
+ * <br>
+ * Theory of operation:
+ * Each search creates a {@link CCSUsage.Builder}, which can be updated during the progress of the search request,
+ * and then it instantiates a {@link CCSUsage} object when the request is finished.
+ * That object is passed to {@link #updateUsage(CCSUsage)} on the request processing end (whether successful or not).
+ * The {@link #updateUsage(CCSUsage)} method will then update the internal counters and metrics.
+ * <br>
+ * When we need to return the current state of the telemetry, we can call {@link #getCCSTelemetrySnapshot()} which produces
+ * a snapshot of the current state of the telemetry as {@link CCSTelemetrySnapshot}. These snapshots are additive so
+ * when collecting the snapshots from multiple nodes, an empty snapshot is created and then all the node's snapshots are added
+ * to it to obtain the summary telemetry.
  */
 public class CCSUsageTelemetry {
 
-    private final LongAdder totalCCSCount;  // TODO: need this? or just sum the successfulSearchTelem and failedSearchTelem ?
-    private final SuccessfulCCSTelemetry successfulSearchTelem;
-    private final FailedCCSTelemetry failedSearchTelem;
+    /**
+     * Result of the request execution.
+     * Either "success" or a failure reason.
+     */
+    public enum Result {
+        SUCCESS("success"),
+        REMOTES_UNAVAILABLE("remotes_unavailable"),
+        CANCELED("canceled"),
+        // TODO: do we need this? If we subtract all known reasons, the rest is unknown.
+        // May be helpful though if there's a lot of other reasons and it may be hard to calculate the unknowns for some clients.
+        UNKNOWN("unknown");
+
+        private final String name;
+
+        Result(String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+    }
+
+    // Not enum because we won't mind other places adding their own features
```

Comment:
I'm not sure what this comment means - can you clarify. Are you saying that more features could be added? If yes, why would using an enum not work?

Reply:
Because of something like async - i.e. say we add a module that does some functionality like some third kind of search, beyond regular and async, and we want to count it. If we make enum, we'd have to make change to this class each time it happens. But if it's just a string then we can make that module just call `setFeature` and this class doesn't need to know.

That said, we don't actually have (yet?) a good way for async or other modules to call `setFeature` - I'd like to make it happen and eliminate the complex async detection code, but I left it for the future for now. So if we wanted the enum right now and switch to a string later we could (I think serialization would work the same?) I just made it a string with an eye to eventually making the solution where async and others take care of marking themselves without telemetry class knowing.